saga-core/src/main/java/io/servicecomb/saga/core/ForwardRecovery.java

```diff
@@ -26,22 +26,25 @@
 
   @Override
   public void apply(SagaTask task, SagaRequest request) {
-    boolean success = false;
-
-    do {
-      try {
-        task.commit(request);
-        success = true;
-      } catch (SagaStartFailedException e) {
-        throw e;
-      } catch (Exception e) {
-        log.error("Applying {} policy due to failure in transaction {} of service {}",
-            description(),
-            request.transaction(),
-            request.serviceName(),
-            e
-        );
-      }
-    } while (!success);
+    try {
+      boolean success = false;
+      do {
+        try {
+          task.commit(request);
+          success = true;
+        } catch (SagaStartFailedException e) {
+          throw e;
+        } catch (Exception e) {
+          log.error("Applying {} policy due to failure in transaction {} of service {}",
+              description(),
+              request.transaction(),
+              request.serviceName(),
+              e
+          );
+          Thread.sleep(request.failRetryDelayMilliseconds());
+        }
+      } while (!success);
+    } catch (InterruptedException ignored) {
+    }
```

Comment:
preferably log a warning in catch block to make troubleshooting easier.

Reply:
Done